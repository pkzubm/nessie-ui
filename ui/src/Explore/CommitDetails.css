/**
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.commitDetails__branchFooter{
    display: flex;
}
.commitDetails__authorFooter{
    display: flex;
    justify-content: space-between;
    background-color: #fff !important;
    word-break: break-word;
    border: 1px solid #d8d8d8;
    border-top: 0px;
}
.commitDetails__authorName{
    font-weight: 600;
}
.commitDetails__authorFooter span{
    font-size: 14px;
}
.commitTables_accordionSummary{
    display: flex;
    flex-direction: row-reverse;
    background-color: #f7f7f7 !important;
}
.commitTables .MuiAccordion-rounded{
    margin-bottom: 10px !important;
}
