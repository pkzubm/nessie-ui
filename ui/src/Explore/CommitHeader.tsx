/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Card } from "react-bootstrap";
import prettyMilliseconds from "pretty-ms";
import React, { useEffect, useState } from "react";
import { api, CommitMeta } from "../utils";
import { factory } from "../ConfigLog4j";

const log = factory.getLogger("api.CommitHeader");

const fetchLog = (
  currentRef: string
): Promise<void | CommitMeta | undefined> => {
  return api()
    .getCommitLog({ ref: currentRef })
    .then((data) => {
      if (data.operations && data.operations.length > 0) {
        return data.operations[0];
      }
    })
    .catch((t) => log.error("CommitLog", t));
};

const CommitHeader = (props: { currentRef: string }): React.ReactElement => {
  const [currentLog, setLog] = useState<CommitMeta>({
    author: undefined,
    authorTime: undefined,
    commitTime: undefined,
    committer: undefined,
    hash: undefined,
    message: "",
    properties: {},
    signedOffBy: undefined,
  });
  useEffect(() => {
    const logs = async () => {
      const results = await fetchLog(props.currentRef);
      if (results) {
        setLog(results);
      }
    };

    void logs();
  }, [props.currentRef]);

  if (!currentLog || !currentLog.hash) {
    return <Card.Header />;
  }
  const properties = Object.keys(currentLog.properties)
    .map((data) => [data, currentLog.properties[data]])
    .map(([k, v]) => `${k}=${v}`)
    .join("; ");
  return (
    <Card.Header>
      <span className={"float-left"}>
        <span className="font-weight-bold">
          {currentLog.committer ?? currentLog.author}
        </span>
        <span>{currentLog.message + " " + properties}</span>
      </span>
      <span className={"float-right"}>
        <span className="font-italic">{currentLog.hash?.slice(0, 8)}</span>
        <span className={"pl-3"}>
          {prettyMilliseconds(
            new Date().getTime() -
              (currentLog.commitTime ?? new Date(0)).getTime(),
            { compact: true }
          )}
        </span>
      </span>
    </Card.Header>
  );
};

export default CommitHeader;
