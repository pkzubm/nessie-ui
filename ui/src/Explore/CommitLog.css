/**
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.commitLog__messageLink {
    color: #000 !important;
    font-weight: 600;
}
.commitLog__messageLink:hover {
    color: #007bff !important;
    text-decoration: underline;
}
.commitLog__body{
    display: flex;
    justify-content: space-between;
    align-items: center;
}
.commitLog__btnWrapper{
    display: flex;
    align-items: center;
}
.commitLog__hashBtnWrapper{
    display: flex;
    align-items: center;
    background-color: #f5f7f9;
}
.commitLog__notFoundText{
    display: flex;
    justify-content: center;
}
.commitLog__copyBtn{
    color: #000 !important;
}
.commitLog__hashBtn{
    color: #0969da !important;
}
.rightBtnHover:hover{
    color: #f5f7f9 !important;
    background-color: #0969da;
}
.commitLog__pagination{
    padding: 10px;
    justify-content: flex-end;
}