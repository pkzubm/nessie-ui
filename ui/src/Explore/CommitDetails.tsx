/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Card } from "react-bootstrap";
import moment from "moment";
import React, { Fragment } from "react";
import { CommitMeta } from "../utils";
import "./CommitDetails.css";
import { EmptyMessageView } from "./Components";

const CommitDetails = (props: {
  commitDetails: CommitMeta | undefined;
  currentRef: string;
}): React.ReactElement => {
  const { commitDetails, currentRef } = props;

  const renderDetailsView = () => {
    if (!commitDetails) return null;
    const {
      message,
      author,
      authorTime,
      signedOffBy,
      properties,
      hash,
      committer,
      commitTime,
    } = commitDetails;
    const additionalProperties = Object.keys(properties)
      .map((data) => [data, properties[data]])
      .map(([k, v]) => `${k}: ${v}`)
      .join(";");
    const propertiesArr = additionalProperties.split(";");

    const authorHoursDiff = moment().diff(moment(authorTime), "hours");
    const commitHoursDiff = moment().diff(moment(commitTime), "hours");

    const authorTimeAgo =
      authorHoursDiff > 24
        ? moment(authorTime).format("YYYY-MM-DD, hh:mm a")
        : `${moment(authorTime).fromNow()}`;
    const commitTimeAgo =
      commitHoursDiff > 24
        ? moment(commitTime).format("YYYY-MM-DD, hh:mm a")
        : `${moment(commitTime).fromNow()}`;

    return (
      <Card.Body className="commitDetails">
        <Card.Header>
          {message && <Card.Title>{message}</Card.Title>}
          {author && (
            <Card.Text className={"ml-3"}>
              <span className={"ml-2"}>Authored by: </span>
              <span className={"ml-2"}>{author}</span>
              <span className={"ml-2"}>at</span>
              <span className={"ml-2"}>{authorTimeAgo}</span>
            </Card.Text>
          )}
          {signedOffBy && (
            <Card.Text className={"ml-3"}>
              <span className={"ml-2"}>Signed off by: </span>
              <span className={"ml-2"}>{signedOffBy}</span>
            </Card.Text>
          )}
          {committer && (
            <Card.Text className={"ml-3"}>
              <span className={"ml-2"}>Committed by: </span>
              <span className={"ml-2"}>{committer}</span>
              <span className={"ml-2"}>at</span>
              <span className={"ml-2"}>{commitTimeAgo}</span>
            </Card.Text>
          )}
          {propertiesArr?.length > 0 && (
            <div className={"ml-3"}>
              <span className={"ml-2"}>Properties: </span>
              <ul>
                {propertiesArr.map((add, index) => {
                  return (
                    <li className={"ml-4"} key={index}>
                      {add}
                    </li>
                  );
                })}
              </ul>
            </div>
          )}
        </Card.Header>
        <Card.Footer className="commitDetails__branchFooter">
          <img
            alt="branchIcon"
            width="22"
            height="22"
            src="/branchIcon.svg"
            className="mr-2"
          />
          <Card.Title>{currentRef}</Card.Title>
        </Card.Footer>
        <Card.Footer className="commitDetails__authorFooter">
          <div>
            <span className="commitDetails__authorName mr-1">{author}</span>
            <span>committed on {moment(commitTime).fromNow()}</span>
          </div>
          <span>commit {hash}</span>
        </Card.Footer>
      </Card.Body>
    );
  };

  if (!commitDetails) {
    return <EmptyMessageView />;
  }

  return <Fragment>{renderDetailsView()}</Fragment>;
};

export default CommitDetails;
