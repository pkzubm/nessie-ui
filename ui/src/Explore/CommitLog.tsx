/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, Card, Nav } from "react-bootstrap";
import moment from "moment";
import React, { useEffect, useState, Fragment } from "react";
import { useParams, useHistory, useLocation } from "react-router-dom";
import { api, CommitMeta } from "../utils";
import { factory } from "../ConfigLog4j";
import CodeIcon from "@material-ui/icons/Code";
import "./CommitLog.css";
import { Icon, Tooltip } from "@material-ui/core";
import ExploreLink from "./ExploreLink";
import { routeSlugs } from "./Constants";
import { EmptyMessageView } from "./Components";
import CommitDetails from "./CommitDetails";

const log = factory.getLogger("api.CommitHeader");

const fetchLog = (ref: string): Promise<void | CommitMeta[] | undefined> => {
  return api()
    .getCommitLog({ ref })
    .then((data) => {
      if (data.operations && data.operations.length > 0) {
        return data.operations;
      }
    })
    .catch((t) => log.error("CommitLog", t));
};

const CommitLog = (props: {
  currentRef: string;
  path: string[];
}): React.ReactElement => {
  const { currentRef, path } = props;
  const { slug } = useParams<{ slug: string }>();
  const [logList, setLogList] = useState<CommitMeta[]>([
    {
      author: undefined,
      authorTime: undefined,
      commitTime: undefined,
      committer: undefined,
      hash: undefined,
      message: "",
      properties: {},
      signedOffBy: undefined,
    },
  ]);
  const [showCommitDetails, setShowCommitDetails] = useState(false);
  const [commitDetails, setCommitDetails] = useState<CommitMeta | undefined>();
  const location = useLocation();
  const history = useHistory();
  useEffect(() => {
    const logs = async () => {
      const results = await fetchLog(currentRef);
      if (results) {
        setLogList(results);
      }
      if (showCommitDetails) {
        const listPath = location.pathname.substring(
          0,
          location.pathname.lastIndexOf("/")
        );
        history.push(listPath);
      }
    };
    void logs();
  }, [currentRef]);

  useEffect(() => {
    if (slug) {
      const last = slug.substring(slug.lastIndexOf("/") + 1, slug.length);
      setShowCommitDetails(last !== routeSlugs.commits);
      const logDetails = logList.find((logItem) => logItem.hash === last);
      setCommitDetails(logDetails);
    }
  }, [slug, logList]);

  if (!logList || (logList.length === 1 && !logList[0].hash)) {
    return <EmptyMessageView />;
  }

  const copyHash = async (hashCode: string) => {
    await navigator.clipboard.writeText(hashCode);
  };
  const commitList = (currentLog: CommitMeta, index: number) => {
    const { commitTime, author, message, hash } = currentLog;
    const hoursDiff = moment().diff(moment(commitTime), "hours");
    const dateTimeAgo =
      hoursDiff > 24
        ? moment(commitTime).format("YYYY-MM-DD, hh:mm a")
        : `${moment(commitTime).fromNow()}`;

    return (
      <Fragment key={index}>
        <Card.Body className="commitLog__body border-bottom">
          <div>
            <Nav.Item>
              <ExploreLink
                toRef={currentRef}
                path={path.concat(hash ?? "#")}
                type="CONTAINER"
                className="commitLog__messageLink"
              >
                {message}
              </ExploreLink>
            </Nav.Item>
            <Card.Text className={"ml-3"}>
              {author}
              <span className={"ml-2"}>committed on</span>
              <span className={"ml-2"}>{dateTimeAgo}</span>
            </Card.Text>
          </div>
          <div className={"commitLog__btnWrapper"}>
            <div className={"border rounded commitLog__hashBtnWrapper"}>
              <Tooltip title="Copy hash">
                <Button
                  className="border-right commitLog__copyBtn rightBtnHover"
                  variant="link"
                  onClick={() => copyHash(hash || "")}
                >
                  <Icon>content_copy</Icon>
                </Button>
              </Tooltip>
              <ExploreLink
                toRef={currentRef}
                path={path.concat(hash ?? "#")}
                type="CONTAINER"
                className="commitLog__hashBtn rightBtnHover"
              >
                <Tooltip title="Commit details">
                  <Button variant="link">
                    <span className="font-italic">{hash?.slice(0, 8)}</span>
                  </Button>
                </Tooltip>
              </ExploreLink>
            </div>
            <div>
              <Tooltip title="Browse the repository at this point in the history">
                <Button variant="light" className={"ml-3 rightBtnHover"}>
                  <CodeIcon />
                </Button>
              </Tooltip>
            </div>
          </div>
        </Card.Body>
      </Fragment>
    );
  };

  return (
    <Card className={"commitLog"}>
      {!showCommitDetails ? (
        logList.map((item, index) => {
          return commitList(item, index);
        })
      ) : (
        <CommitDetails commitDetails={commitDetails} currentRef={currentRef} />
      )}
    </Card>
  );
};

export default CommitLog;
