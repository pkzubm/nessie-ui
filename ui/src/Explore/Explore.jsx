/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, {useEffect, useState} from 'react';
import {Card, Container} from "react-bootstrap";
import {useParams} from "react-router-dom";
import TableHead from "./TableHead";
import TableListing from "./TableListing";
import CommitLog from "./CommitHeader";
import {api} from "../utils";


function fetchDefaultBranch(setDefaultBranch) {
  return api().getDefaultBranch()
    .then(res => {
      return res.json();
    })
    .then((data) => {
      setDefaultBranch(data.name);
    }).catch(t => console.log(t))
}

function btCompare(a, b) {
  if (a.name < b.name) {
    return -1;
  }
  if (a.name > b.name) {
    return 1;
  }

  return 0;
}

function loadBranchesAndTags(setBranches, setTags) {
  return api().getAllReferences()
    .then(res => {
      let data = res.json();
      return data;
    })
    .then((data) => {
      data = data.sort(btCompare);
      setBranches(data.filter(x => x.type === "BRANCH"));
      setTags(data.filter(x => x.type === "TAG"));
    });
}

function parseSlug(slug, defaultBranch, branches, tags) {

  if (!slug || slug.split("/").length === 0) {
    return {currentRef: defaultBranch, path: []};
  }

  // if we have a slug, need to figure out what portion is related to a ref versus a key.
  let sub = "";
  let checker = b => {

    if (b.name.length <= sub.length) {
      return;
    }
    if (slug.toLowerCase().startsWith(b.name.toLowerCase())) {
      sub = b.name;
    }
  };

  branches.forEach(checker);
  tags.forEach(checker);

  if (sub.length === 0) {
    sub = slug.split("/")[0];
  }

  let path = slug.slice(sub.length).split("/").filter(i => i);

  return {currentRef: sub, path: path};
}

function updateRef(slug, defaultBranch, branches, tags, setPath, setCurrentRef) {
  if (!defaultBranch || !branches) {
    return;
  }

  const {currentRef, path} = parseSlug(slug, defaultBranch, branches, tags);
  setCurrentRef(currentRef);
  setPath(path);
}

function Explore(props) {

  let { slug } = useParams();
  const [defaultBranch, setDefaultBranch] = useState("main");
  const [branches, setBranches] = useState([]);
  const [tags, setTags] = useState([]);
  const [path, setPath] = useState([]);
  const [currentRef, setCurrentRef] = useState();

  useEffect(() => {
    fetchDefaultBranch(setDefaultBranch);
    loadBranchesAndTags(setBranches, setTags);
  }, []);

  useEffect(() => {
    updateRef(slug, defaultBranch, branches, tags, setPath, setCurrentRef);
  }, [slug, defaultBranch, branches, tags]);

  return (
    <div>
      <Container style={{"marginTop": "100px"}}>
        <Card>
          <TableHead
            branches={branches}
            tags={tags}
            currentRef={currentRef}
            defaultBranch={defaultBranch}
            path={path}
          />
          <CommitLog currentRef={currentRef} />
          <TableListing currentRef={currentRef} path={path} />
        </Card>
      </Container>
    </div>
  );
}

Explore.propTypes = {}

export default Explore;
