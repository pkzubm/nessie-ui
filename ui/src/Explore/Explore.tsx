/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { useEffect, useState } from "react";
import { Card, Container } from "react-bootstrap";
import { useParams } from "react-router-dom";
import TableHead from "./TableHead";
import TableListing from "./TableListing";
import { routeSlugs } from "./Constants";
import CommitLog from "./CommitLog";
import { api, Branch, Reference, Tag } from "../utils";
import { factory } from "../ConfigLog4j";
import "./Explore.css";

const log = factory.getLogger("api.Explore");

const fetchDefaultBranch = (): Promise<string | void> => {
  return api()
    .getDefaultBranch()
    .then((data) => {
      if (data.name) {
        return data.name;
      }
    })
    .catch((t) => log.error("DefaultBranch", t as undefined));
};

const loadBranchesAndTags = (): Promise<Reference[] | void> => {
  return api()
    .getAllReferences({})
    .then((data) => {
      return data.references.sort(btCompare);
    })
    .catch((t) => log.error("AllReferences", t as undefined));
};

const btCompare = (a: Branch, b: Branch): number => {
  if (!a.name) {
    return !b.name ? 0 : 1;
  } else if (!b.name) {
    return -1;
  }

  if (a.name < b.name) {
    return -1;
  }
  if (a.name > b.name) {
    return 1;
  }

  return 0;
};

interface Slug {
  currentRef: string;
  path: string[];
}

const parseSlug = (
  slug: string,
  defaultBranch: string,
  branches: Branch[],
  tags: Tag[]
): Slug => {
  if (!slug || slug.split("/").length === 0) {
    return { currentRef: defaultBranch, path: [] };
  }

  // if we have a slug, need to figure out what portion is related to a ref versus a key.
  let sub = "";
  const checker = (b: Branch) => {
    if (!b.name) {
      return;
    }
    if (b.name.length <= sub.length) {
      return;
    }
    if (slug.toLowerCase().startsWith(b.name.toLowerCase())) {
      sub = b.name ?? "";
    }
  };

  branches.forEach(checker);
  tags.forEach(checker);

  if (sub.length === 0) {
    sub = slug.split("/")[0];
  }

  const path = slug
    .slice(sub.length)
    .split("/")
    .filter((i) => i);

  return { currentRef: sub, path };
};

const updateRef = (
  slug: string,
  defaultBranch: string,
  branches: Branch[],
  tags: Tag[]
): Slug | void => {
  if (!defaultBranch || !branches) {
    return;
  }

  return parseSlug(slug, defaultBranch, branches, tags);
};

interface IBranchState {
  branches: Branch[];
  tags: Tag[];
}

const Explore = (): React.ReactElement => {
  const { slug } = useParams<{ slug: string }>();
  const [defaultBranch, setDefaultBranch] = useState<string>("main");
  const [branches, setBranches] = useState<IBranchState>({
    branches: [],
    tags: [],
  });
  const [path, setPath] = useState<string[]>([]);
  const [currentRef, setCurrentRef] = useState<string>();

  useEffect(() => {
    const references = async () => {
      const results = await loadBranchesAndTags();
      if (results) {
        setBranches({
          branches: results
            .filter((x) => x.type === "BRANCH")
            .map((b) => b as Branch),
          tags: results.filter((x) => x.type === "TAG").map((t) => t as Tag),
        });
      }
    };
    const getDefaultBranch = async () => {
      const results = await fetchDefaultBranch();
      if (results) {
        setDefaultBranch(results);
      }
    };
    void references();
    void getDefaultBranch();
  }, []);

  useEffect(() => {
    if (branches.branches.length === 0 && branches.tags.length === 0) {
      return;
    }
    const newSlug = updateRef(
      slug,
      defaultBranch,
      branches.branches,
      branches.tags
    );
    if (newSlug) {
      setCurrentRef(newSlug.currentRef);
      setPath(newSlug.path);
    }
  }, [slug, defaultBranch, branches]);

  return currentRef ? (
    <div>
      <Container className="explore">
        <Card className="explore__wrapper">
          <TableHead
            branches={branches.branches}
            tags={branches.tags}
            currentRef={currentRef}
            defaultBranch={defaultBranch}
            path={path}
          />
          {path.length > 0 && path[0] === routeSlugs.commits ? (
            <CommitLog currentRef={currentRef} path={path} />
          ) : (
            <TableListing
              currentRef={currentRef}
              path={path}
              branches={branches.branches}
            />
          )}
        </Card>
      </Container>
    </div>
  ) : (
    <div>
      <Container style={{ marginTop: "100px" }}>
        <Card />
      </Container>
    </div>
  );
};

Explore.propTypes = {};

export default Explore;
