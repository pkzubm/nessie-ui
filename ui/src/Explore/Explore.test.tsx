/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { BrowserRouter, Route, Router } from "react-router-dom";
import { render, screen, waitFor } from "@testing-library/react";
import Explore from "./Explore";
// tslint:disable-next-line:no-implicit-dependencies
import nock from "nock";
import { createMemoryHistory } from "history";

// noinspection TypeScriptValidateTypes
const renderWithRouter = (component: React.ReactElement) => {
  const history = createMemoryHistory({
    initialEntries: ["/tree/main/a"],
  });
  // eslint-disable-next-line react/prop-types
  const Wrapper = ({ children }) => (
    <Router history={history}>
      <Route path="/tree/:slug">{children}</Route>
    </Router>
  );
  return {
    ...render(component, { wrapper: Wrapper }),
    history,
  };
};

it("Explore renders", async () => {
  const entry = {
    name: { elements: ["a"] },
    type: "UNKNOWN",
  };
  const now = new Date();
  now.setDate(now.getDate() - 1);
  const commitMeta = {
    hash: "deadbeef",
    author: "bob",
    commitTime: now,
    committer: "sally",
    message: "commitMessage",
    properties: { a: "b", c: "d" },
  };
  const scope1 = nock("http://localhost/api/v1")
    .get("/trees/tree/main/log")
    .reply(200, { token: "foo", operations: [commitMeta] });
  const scope2 = nock("http://localhost/api/v1")
    .get(
      "/trees/tree/main/entries?namespaceDepth=1&query_expression=entry.namespace.matches(%27(%5C%5C.%7C%24)%27)"
    )
    .reply(200, { token: "foo", entries: [entry] });
  const scope3 = nock("http://localhost/api/v1")
    .get("/trees/tree")
    .reply(200, { name: "main", hash: "deadbeef" });
  const scope4 = nock("http://localhost/api/v1")
    .get("/trees")
    .reply(200, [tag("a"), tag("b"), branch("c"), branch("d")]);

  const { getByText, asFragment } = render(
    <React.StrictMode>
      <BrowserRouter>
        <Explore />
      </BrowserRouter>
    </React.StrictMode>
  );
  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => getByText("a"));

  scope1.done();
  scope2.done();
  scope3.done();
  scope4.done();
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByText("a").closest("a")).toHaveAttribute(
    "href",
    "/tree/main/a"
  );
});

it("Explore renders with slug", async () => {
  const entry = {
    name: { elements: ["a"] },
    type: "UNKNOWN",
  };
  const now = new Date();
  now.setDate(now.getDate() - 1);
  const commitMeta = {
    hash: "deadbeef",
    author: "bob",
    commitTime: now,
    committer: "sally",
    message: "commitMessage",
    properties: { a: "b", c: "d" },
  };
  const scope1 = nock("http://localhost/api/v1")
    .get("/trees/tree/main/log")
    .reply(200, { token: "foo", operations: [commitMeta] });
  const scope2 = nock("http://localhost/api/v1")
    .get(
      "/trees/tree/main/entries?namespaceDepth=1&query_expression=entry.namespace.matches(%27(%5C%5C.%7C%24)%27)"
    )
    .reply(200, { token: "foo", entries: [entry] });
  const scope3 = nock("http://localhost/api/v1")
    .get("/trees/tree")
    .reply(200, { name: "main", hash: "deadbeef" });
  const scope4 = nock("http://localhost/api/v1")
    .get("/trees")
    .reply(200, [tag("a"), tag("b"), branch("c"), branch("d")]);
  // jest.spyOn(MemoryRouter, 'useParams').mockReturnValue({ slug: "/main/a" });

  const { getByText, asFragment } = renderWithRouter(<Explore />);
  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => getByText("a"));

  scope1.done();
  scope2.done();
  scope3.done();
  scope4.done();
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByText("a").closest("a")).toHaveAttribute(
    "href",
    "/tree/main/a"
  );
});

const tag = (name: string, hash = "deadbeef") => {
  return { name, hash, type: "TAG" };
};

const branch = (name: string, hash = "deadbeef") => {
  return { name, hash, type: "BRANCH" };
};
