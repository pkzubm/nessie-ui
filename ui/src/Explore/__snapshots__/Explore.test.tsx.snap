// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explore renders 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
      style="margin-top: 100px;"
    >
      <div
        class="card"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Explore renders 2`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
      style="margin-top: 100px;"
    >
      <div
        class="card"
      >
        <div
          class="nav card-header-pills nav-pills"
        >
          <div
            class="nav-item dropdown"
            style="padding-left: 1em;"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle nav-link"
              href="#"
              id="nav-dropdown"
              role="button"
            >
              main
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 16l-4-4V8.82C14.16 8.4 15 7.3 15 6c0-1.66-1.34-3-3-3S9 4.34 9 6c0 1.3.84 2.4 2 2.82V12l-4 4H3v5h5v-3.05l4-4.2 4 4.2V21h5v-5h-4z"
                />
              </svg>
              2
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M21.41 11.58l-9-9C12.05 2.22 11.55 2 11 2H4c-1.1 0-2 .9-2 2v7c0 .55.22 1.05.59 1.42l9 9c.36.36.86.58 1.41.58.55 0 1.05-.22 1.41-.59l7-7c.37-.36.59-.86.59-1.41 0-.55-.23-1.06-.59-1.42zM5.5 7C4.67 7 4 6.33 4 5.5S4.67 4 5.5 4 7 4.67 7 5.5 6.33 7 5.5 7z"
                />
              </svg>
              2
            </a>
          </div>
        </div>
        <div
          class="card-header"
        >
          <span
            class="float-left"
          >
            <span
              class="font-weight-bold"
            >
              sally
            </span>
            <span>
              commitMessage a=b; c=d
            </span>
          </span>
          <span
            class="float-right"
          >
            <span
              class="font-italic"
            >
              deadbeef
            </span>
            <span
              class="pl-3"
            >
              1d
            </span>
          </span>
        </div>
        <div
          class="card"
        >
          <div
            class="list-group list-group-flush"
          >
            <div
              class="list-group-item"
            >
              <a
                href="/tree/main/a"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
                  />
                </svg>
                a
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Explore renders with slug 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
      style="margin-top: 100px;"
    >
      <div
        class="card"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Explore renders with slug 2`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
      style="margin-top: 100px;"
    >
      <div
        class="card"
      >
        <div
          class="nav card-header-pills nav-pills"
        >
          <div
            class="nav-item dropdown"
            style="padding-left: 1em;"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle nav-link"
              href="#"
              id="nav-dropdown"
              role="button"
            >
              main
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 16l-4-4V8.82C14.16 8.4 15 7.3 15 6c0-1.66-1.34-3-3-3S9 4.34 9 6c0 1.3.84 2.4 2 2.82V12l-4 4H3v5h5v-3.05l4-4.2 4 4.2V21h5v-5h-4z"
                />
              </svg>
              2
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M21.41 11.58l-9-9C12.05 2.22 11.55 2 11 2H4c-1.1 0-2 .9-2 2v7c0 .55.22 1.05.59 1.42l9 9c.36.36.86.58 1.41.58.55 0 1.05-.22 1.41-.59l7-7c.37-.36.59-.86.59-1.41 0-.55-.23-1.06-.59-1.42zM5.5 7C4.67 7 4 6.33 4 5.5S4.67 4 5.5 4 7 4.67 7 5.5 6.33 7 5.5 7z"
                />
              </svg>
              2
            </a>
          </div>
        </div>
        <div
          class="card-header"
        >
          <span
            class="float-left"
          >
            <span
              class="font-weight-bold"
            >
              sally
            </span>
            <span>
              commitMessage a=b; c=d
            </span>
          </span>
          <span
            class="float-right"
          >
            <span
              class="font-italic"
            >
              deadbeef
            </span>
            <span
              class="pl-3"
            >
              1d
            </span>
          </span>
        </div>
        <div
          class="card"
        >
          <div
            class="list-group list-group-flush"
          >
            <div
              class="list-group-item"
            >
              <a
                href="/tree/main/a"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
                  />
                </svg>
                a
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
